src/averageAges.js

```diff
@@ -1,64 +1,46 @@
 'use strict';
 
-/**
- * Implement calculateMenAverageAge function
- *
- * Function returns average age of men in array. If `century` is specified then
- * function calculates average age only for men who died in this century
- *
- * To calculate century:
- * Divide year of person's death by 100: Math.ceil(person.died / 100)
- *
- * @param {object[]} people
- * @param {number} century - optional
- *
- * @return {number}
- */
 function calculateMenAverageAge(people, century) {
-  // write code here
-  // learn how to use array methods like .filter .map .some .every .find .reduce
-  // avoid using loop and forEach
-  // replace `if ()` statement with &&, || or ?:
-  // without nesting
+  const onlyMen = genderFilter(people, 'm');
+
+  const filteredMen = (century)
+    ? onlyMen.filter(person => Math.ceil(person.died / 100) === century)
+    : onlyMen;
+
+  return calculateAvgAge(filteredMen);
 }
 
-/**
- * Implement calculateWomenAverageAge function
- *
- * Function returns average age of women in array. If `withChildren` is
- * specified then function calculates average age only for women with children
- *
- * Hint: To check if a woman has children you should find someone who mention
- * her as mother.
- *
- * @param {object[]} people
- * @param {boolean} withChildren - optional
- *
- * @return {number}
- */
 function calculateWomenAverageAge(people, withChildren) {
-  // write code here
+  const onlyWomen = genderFilter(people, 'f');
+
+  const filteredWomen = (withChildren)
+    ? onlyWomen.filter(woman => people.some(p => p.mother === woman.name))
+    : onlyWomen;
+
+  return calculateAvgAge(filteredWomen);
 }
 
-/**
- * Implement calculateAverageAgeDiff function.
- *
- * The function returns an average age difference between a child and his or her
- * mother in the array. (A mother's age at child birth)
- *
- * If `onlyWithSon` is specified then function calculates age difference only
- * for sons and their mothers.
- *
- * @param {object[]} people
- * @param {boolean} onlyWithSon - optional
- *
- * @return {number}
- */
 function calculateAverageAgeDiff(people, onlyWithSon) {
-  // 1. find a mother of each person (or only for men)
-  // 2. keep people who have mothers in the array
-  // 3. calculate the difference child.born - mother.born
-  // 4. return the average value
+  const filteredChildren = people.filter(person => (onlyWithSon)
+    ? people.find(mother => mother.name === person.mother
+      && person.sex === 'm')
+    : people.find(mother => mother.name === person.mother)
+  );
+
+  const ageDiff = filteredChildren.map(child =>
+    child.born - people.find(mother => mother.name === child.mother).born);
```

Comment:
<img width="1046" alt="image" src="https://github.com/mate-academy/js_average-ages/assets/86789161/e1f951c7-20ca-4e2f-83d3-81ced65b01bc">

But better will be not to write it in one line. Better will be to create variable for `people.find(mother => mother.name === child.mother).born`
Also think about using destructuring

Reply:
i tried to put it in a variable but then it wouldn't work as expected and some tests would fail so I did not do it 🙃